src/fides/api/alembic/migrations/versions/fd52d5f08c17_system_dictionary_data_migration.py

```diff
@@ -0,0 +1,374 @@
+
+
+def _system_transform_data_responsibility_title_type(
+    dept_str: Optional[str],
+) -> List[str]:
+    """Migrating System.data_responsibility_title (str) to System.responsibility(list).  There can now be potentially many
+    items listed here"""
+    if not dept_str:
+        return []
+    return [dept_str]
+
+
+def _flatten_data_protection_impact_assessment(
+    data_protection_impact_assessment: Optional[Dict],
+) -> Tuple[bool, Optional[str], Optional[str]]:
+    """Flattening data from System.data_protection_impact_assessment into three separate fields:
+    System.requires_data_protection_assessments, System.dpa_location, and System.dpa_progress
+    """
+    if not data_protection_impact_assessment or not isinstance(
+        data_protection_impact_assessment, dict
+    ):
+        return False, None, None
+
+    requires_data_protection_assessments = data_protection_impact_assessment.get(
+        "is_required", False
+    )
+    dpa_progress = data_protection_impact_assessment.get("progress", None)
+    dpa_location = data_protection_impact_assessment.get("link", None)
+    return requires_data_protection_assessments, dpa_progress, dpa_location
+
+
+def system_dictionary_additive_migration(bind: Connection):
+    """Copy and/or transform existing data in Datasets, PrivacyDeclarations and the System itself to new
+    Systems fields to support dictionary work"""
+    get_systems_query: TextClause = text(
+        """
+        SELECT 
+            ctl_systems.id,
+            pgp_sym_decrypt(ctl_systems.joint_controller::bytea, :encryption_key) AS joint_controller_decrypted,
+            ctl_systems.third_country_transfers,
+            ctl_systems.data_responsibility_title,
+            ctl_systems.data_protection_impact_assessment,
+            array_accum(ctl_datasets.third_country_transfers) as dataset_third_country_transfers,
+            array_agg(pgp_sym_decrypt(ctl_datasets.joint_controller::bytea, :encryption_key)) AS datasets_joint_controller_decrypted
+        FROM ctl_systems
+        LEFT JOIN privacydeclaration ON ctl_systems.id = privacydeclaration.system_id 
+        LEFT JOIN ctl_datasets ON ctl_datasets.fides_key = ANY(privacydeclaration.dataset_references) 
+        GROUP BY ctl_systems.id
+        """
+    )
+    for row in bind.execute(
+        get_systems_query, {"encryption_key": CONFIG.user.encryption_key}
+    ):
+        (
+            requires_data_protection_assessments,
+            dpa_progress,
+            dpa_location,
+        ) = _flatten_data_protection_impact_assessment(
+            row["data_protection_impact_assessment"]
+        )
+
+        system_data = {
+            JOINT_CONTROLLER_INFO: _system_transform_joint_controller_info(
+                json.loads(row["joint_controller_decrypted"] or "null"),
+                [  # Potentially multiple datasets on systems that have joint controller info so this is a list
```

Comment:
This feels like a lot of nesting to me and obfuscates the function calls in my opinion

Maybe assigning these to variables before passing them would be more readable?

Reply:
yes, good callout!